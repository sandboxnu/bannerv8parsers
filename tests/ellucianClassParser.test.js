/*
 * This file is part of Search NEU and licensed under AGPL3.
 * See the license file in the root folder for details.
 */

import path from 'path';
import MockDate from 'mockdate';
import fs from 'fs-extra';

import ellucianClassParser from '../ellucianClassParser';


beforeAll(() => {
  MockDate.set('Mon Oct 10 2016 00:00:00 -0000', -3000);
});

afterAll(() => {
  MockDate.reset();
});


it('can parse crns', async (done) => {
  const body = await fs.readFile(path.join(__dirname, 'data', 'ellucianClassParser', '1.html'), 'utf8');

  //set up variables
  const url = 'https://wl11gp.neu.edu/udcprod8/bwckctlg.p_disp_listcrse?term_in=201610&subj_in=EECE&crse_in=2160&schd_in=LEC';

  //main parse
  const output = ellucianClassParser.parse(body, url);

  expect(ellucianClassParser.supportsPage(url)).toBe(true);

  const aClass = output.classWrapper.value;

  expect(aClass).toEqual({
    classAttributes: [], crns: ['15633', '15636', '15639', '16102', '17799', '17800'],
  });

  done();
});

it('should parse a bunch of deps', async (done) => {
  //sections have different names
  const body = await fs.readFile(path.join(__dirname, 'data', 'ellucianClassParser', 'multiname.html'), 'utf8');

  //set up variables -- ellucianClassParser url might not be correct
  const url = 'https://myswat.swarthmore.edu/pls/bwckctlg.p_disp_listcrse?term_in=201502&subj_in=PHYS&crse_in=013&schd_in=LE';

  //main parse
  const output = ellucianClassParser.parse(body, url);
  expect(output).toMatchSnapshot();

  const aClass = output.classWrapper.value;

  expect(ellucianClassParser.supportsPage(url)).toBe(true);

  expect(aClass).toEqual({
    classAttributes: [],
    crns: [
      '24600',
      '24601',
      '24603',
      '25363',
    ],
  });

  done();
});

it('should parse a bunch of crns', async (done) => {
  const body = await fs.readFile(path.join(__dirname, 'data', 'ellucianClassParser', '3.html'), 'utf8');
  //set up variables
  const url = 'https://prd-wlssb.temple.edu/prod8/bwckctlg.p_disp_listcrse?term_in=201503&subj_in=ACCT&crse_in=2102&schd_in=BAS';


  //main parse
  const output = ellucianClassParser.parse(body, url);

  const aClass = output.classWrapper.value;

  expect(ellucianClassParser.supportsPage(url)).toBe(true);

  expect(aClass).toEqual({
    classAttributes: [], crns: ['11018', '11019', '11020', '11679', '19962', '20800', '22497', '23294', '23295', '24435', '6073', '6074', '6075', '6077', '6129', '6130', '8145'],
  });

  expect(output).toMatchSnapshot();

  done();
});


it('should parse a bunch of meetings', async (done) => {
  //lots of different meetings
  const body = await fs.readFile(path.join(__dirname, 'data', 'ellucianClassParser', '4.html'), 'utf8');

  //set up variables
  const url = 'https://prd-wlssb.temple.edu/prod8/bwckctlg.p_disp_listcrse?term_in=201503&subj_in=AIRF&crse_in=2041&schd_in=BAS';

  //main parse
  const output = ellucianClassParser.parse(body, url);

  const aClass = output.classWrapper.value;

  expect(ellucianClassParser.supportsPage(url)).toBe(true);

  expect(aClass).toEqual({
    classAttributes: [],
    crns: ['12090'],
  });

  expect(output).toMatchSnapshot();

  done();
});

it('can parse CANCELLED', async (done) => {
  //cancelled - something was weird with ellucianClassParser one not sure what it was
  const body = await fs.readFile(path.join(__dirname, 'data', 'ellucianClassParser', '6.html'), 'utf8');

  //set up variables
  const url = 'https://ssb.ccsu.edu/pls/ssb_cPROD/bwckctlg.p_disp_listcrse?term_in=201610&subj_in=ANTH&crse_in=245&schd_in=LE';

  //main parse
  const output = ellucianClassParser.parse(body, url);

  const aClass = output.classWrapper.value;

  expect(ellucianClassParser.supportsPage(url)).toBe(true);

  expect(aClass.crns.length).toBe(1);
  expect(aClass.crns[0]).toBe('12291');

  done();
});


//make sure ellucianClassParser.classNameTranslation works
it('name translation works', async (done) => {
  const body = await fs.readFile(path.join(__dirname, 'data', 'ellucianClassParser', 'rename.html'), 'utf8');

  //set up variables
  const url = 'https://wl11gp.neu.edu/udcprod8/bwckctlg.p_disp_listcrse?term_in=201610&subj_in=CS&crse_in=2500&schd_in=LEC';

  //main parse
  const output = ellucianClassParser.parse(body, url);

  // Five sections of fundies, and 1 alt class (hon, which has 1 section).
  expect(output.classWrapper).toMatchSnapshot();

  done();
});
