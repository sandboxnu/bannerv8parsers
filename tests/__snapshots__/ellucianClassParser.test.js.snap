// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`name translation works 1`] = `
Object {
  "deps": Array [],
  "type": "classes",
  "value": Object {
    "classAttributes": Array [],
    "crns": Array [
      "10663",
      "10664",
      "11892",
      "13757",
      "14492",
      "17348",
    ],
  },
}
`;

exports[`should parse a bunch of crns 1`] = `
Object {
  "classWrapper": Object {
    "deps": Array [],
    "type": "classes",
    "value": Object {
      "classAttributes": Array [],
      "crns": Array [
        "11018",
        "11019",
        "11020",
        "11679",
        "19962",
        "20800",
        "22497",
        "23294",
        "23295",
        "24435",
        "6073",
        "6074",
        "6075",
        "6077",
        "6129",
        "6130",
        "8145",
      ],
    },
  },
  "sectionStartingDatas": Array [
    Object {
      "crn": "11018",
      "info": "Managerial Accounting",
      "meetings": Array [
        Object {
          "endDate": 16561,
          "profs": Array [
            "Katharine Wilson",
          ],
          "startDate": 16447,
          "times": Object {
            "1": Array [
              Object {
                "end": 42600,
                "start": 39600,
              },
            ],
            "3": Array [
              Object {
                "end": 42600,
                "start": 39600,
              },
            ],
            "5": Array [
              Object {
                "end": 42600,
                "start": 39600,
              },
            ],
          },
          "type": "Class",
          "where": "Alter Hall 0A231",
        },
      ],
      "url": "https://prd-wlssb.temple.edu/prod8/bwckschd.p_disp_detail_sched?term_in=201503&crn_in=11018",
    },
    Object {
      "crn": "11019",
      "info": "Managerial Accounting",
      "meetings": Array [
        Object {
          "endDate": 16561,
          "profs": Array [
            "Jang Wook Lee",
          ],
          "startDate": 16447,
          "times": Object {
            "1": Array [
              Object {
                "end": 42600,
                "start": 39600,
              },
            ],
            "3": Array [
              Object {
                "end": 42600,
                "start": 39600,
              },
            ],
            "5": Array [
              Object {
                "end": 42600,
                "start": 39600,
              },
            ],
          },
          "type": "Class",
          "where": "Anderson Hall 00027",
        },
      ],
      "url": "https://prd-wlssb.temple.edu/prod8/bwckschd.p_disp_detail_sched?term_in=201503&crn_in=11019",
    },
    Object {
      "crn": "8145",
      "info": "Managerial Accounting",
      "meetings": Array [
        Object {
          "endDate": 16561,
          "profs": Array [
            "Katharine Wilson",
          ],
          "startDate": 16447,
          "times": Object {
            "1": Array [
              Object {
                "end": 35400,
                "start": 32400,
              },
            ],
            "3": Array [
              Object {
                "end": 35400,
                "start": 32400,
              },
            ],
            "5": Array [
              Object {
                "end": 35400,
                "start": 32400,
              },
            ],
          },
          "type": "Class",
          "where": "Alter Hall 0A237",
        },
      ],
      "url": "https://prd-wlssb.temple.edu/prod8/bwckschd.p_disp_detail_sched?term_in=201503&crn_in=8145",
    },
    Object {
      "crn": "6073",
      "info": "Managerial Accounting",
      "meetings": Array [
        Object {
          "endDate": 16561,
          "profs": Array [
            "Carl D. Shelow",
          ],
          "startDate": 16447,
          "times": Object {
            "3": Array [
              Object {
                "end": 72000,
                "start": 63000,
              },
            ],
          },
          "type": "Class",
          "where": "Tuttleman Learning Center 0005B",
        },
      ],
      "url": "https://prd-wlssb.temple.edu/prod8/bwckschd.p_disp_detail_sched?term_in=201503&crn_in=6073",
    },
    Object {
      "crn": "11020",
      "info": "Managerial Accounting",
      "meetings": Array [
        Object {
          "endDate": 16561,
          "profs": Array [
            "Marco Malandra",
          ],
          "startDate": 16447,
          "times": Object {
            "2": Array [
              Object {
                "end": 72000,
                "start": 63000,
              },
            ],
          },
          "type": "Class",
          "where": "Alter Hall 0A032",
        },
      ],
      "url": "https://prd-wlssb.temple.edu/prod8/bwckschd.p_disp_detail_sched?term_in=201503&crn_in=11020",
    },
    Object {
      "crn": "6129",
      "info": "Managerial Accounting",
      "meetings": Array [
        Object {
          "endDate": 16561,
          "profs": Array [
            "Lucas Threinen",
          ],
          "startDate": 16447,
          "times": Object {
            "4": Array [
              Object {
                "end": 72000,
                "start": 63000,
              },
            ],
          },
          "type": "Class",
          "where": "Alter Hall 0A237",
        },
      ],
      "url": "https://prd-wlssb.temple.edu/prod8/bwckschd.p_disp_detail_sched?term_in=201503&crn_in=6129",
    },
    Object {
      "crn": "20800",
      "info": "Managerial Accounting",
      "meetings": Array [
        Object {
          "endDate": 16561,
          "profs": Array [
            "David Heier",
          ],
          "startDate": 16447,
          "times": Object {
            "4": Array [
              Object {
                "end": 72000,
                "start": 63000,
              },
            ],
          },
          "type": "Class",
          "where": "Speakman Hall 00112",
        },
      ],
      "url": "https://prd-wlssb.temple.edu/prod8/bwckschd.p_disp_detail_sched?term_in=201503&crn_in=20800",
    },
    Object {
      "crn": "6074",
      "info": "Managerial Accounting",
      "meetings": Array [
        Object {
          "endDate": 16561,
          "profs": Array [
            "Pamela Lauren Kuperstein",
          ],
          "startDate": 16447,
          "times": Object {
            "3": Array [
              Object {
                "end": 72000,
                "start": 63000,
              },
            ],
          },
          "type": "Class",
          "where": "Alter Hall 0A032",
        },
      ],
      "url": "https://prd-wlssb.temple.edu/prod8/bwckschd.p_disp_detail_sched?term_in=201503&crn_in=6074",
    },
    Object {
      "crn": "23294",
      "info": "Managerial Accounting",
      "meetings": Array [
        Object {
          "endDate": 16561,
          "profs": Array [
            "Dmitri Byzalov",
          ],
          "startDate": 16447,
          "times": Object {
            "1": Array [
              Object {
                "end": 39000,
                "start": 36000,
              },
            ],
            "3": Array [
              Object {
                "end": 39000,
                "start": 36000,
              },
            ],
            "5": Array [
              Object {
                "end": 39000,
                "start": 36000,
              },
            ],
          },
          "type": "Class",
          "where": "Speakman Hall 00112",
        },
      ],
      "url": "https://prd-wlssb.temple.edu/prod8/bwckschd.p_disp_detail_sched?term_in=201503&crn_in=23294",
    },
    Object {
      "crn": "23295",
      "info": "Managerial Accounting",
      "meetings": Array [
        Object {
          "endDate": 16561,
          "profs": Array [
            "Chunnan Zhang",
          ],
          "startDate": 16447,
          "times": Object {
            "1": Array [
              Object {
                "end": 35400,
                "start": 32400,
              },
            ],
            "3": Array [
              Object {
                "end": 35400,
                "start": 32400,
              },
            ],
            "5": Array [
              Object {
                "end": 35400,
                "start": 32400,
              },
            ],
          },
          "type": "Class",
          "where": "Alter Hall 0A231",
        },
      ],
      "url": "https://prd-wlssb.temple.edu/prod8/bwckschd.p_disp_detail_sched?term_in=201503&crn_in=23295",
    },
    Object {
      "crn": "6075",
      "info": "Managerial Accounting",
      "meetings": Array [
        Object {
          "endDate": 16561,
          "profs": Array [
            "Lucas Threinen",
          ],
          "startDate": 16447,
          "times": Object {
            "1": Array [
              Object {
                "end": 72000,
                "start": 63000,
              },
            ],
          },
          "type": "Class",
          "where": "Alter Hall 0A034",
        },
      ],
      "url": "https://prd-wlssb.temple.edu/prod8/bwckschd.p_disp_detail_sched?term_in=201503&crn_in=6075",
    },
    Object {
      "crn": "6077",
      "info": "Managerial Accounting",
      "meetings": Array [
        Object {
          "endDate": 16561,
          "profs": Array [
            "Marco Malandra",
          ],
          "startDate": 16447,
          "times": Object {
            "4": Array [
              Object {
                "end": 72000,
                "start": 63000,
              },
            ],
          },
          "type": "Class",
          "where": "Alter Hall 0A034",
        },
      ],
      "url": "https://prd-wlssb.temple.edu/prod8/bwckschd.p_disp_detail_sched?term_in=201503&crn_in=6077",
    },
    Object {
      "crn": "6130",
      "info": "Managerial Accounting",
      "meetings": Array [
        Object {
          "endDate": 16561,
          "profs": Array [
            "Tom D Adams",
          ],
          "startDate": 16447,
          "times": Object {
            "2": Array [
              Object {
                "end": 39000,
                "start": 34200,
              },
            ],
            "4": Array [
              Object {
                "end": 39000,
                "start": 34200,
              },
            ],
          },
          "type": "Class",
          "where": "Ambler - Learning Center 00306",
        },
      ],
      "url": "https://prd-wlssb.temple.edu/prod8/bwckschd.p_disp_detail_sched?term_in=201503&crn_in=6130",
    },
    Object {
      "crn": "11679",
      "info": "Managerial Accounting",
      "meetings": Array [
        Object {
          "endDate": 16561,
          "profs": Array [
            "Han-Up Park",
          ],
          "startDate": 16447,
          "times": Object {
            "1": Array [
              Object {
                "end": 72000,
                "start": 63000,
              },
            ],
          },
          "type": "Class",
          "where": "TUCC 1515 Market Street 00422",
        },
      ],
      "url": "https://prd-wlssb.temple.edu/prod8/bwckschd.p_disp_detail_sched?term_in=201503&crn_in=11679",
    },
    Object {
      "crn": "22497",
      "info": "Managerial Accounting",
      "meetings": Array [
        Object {
          "endDate": 16561,
          "profs": Array [
            "Christian E. Wurst",
          ],
          "startDate": 16447,
          "times": Object {
            "1": Array [
              Object {
                "end": 70200,
                "start": 61200,
              },
            ],
          },
          "type": "OLL with reqd virtual meetings",
          "where": "Online Room - Place Holder CLSRM",
        },
      ],
      "url": "https://prd-wlssb.temple.edu/prod8/bwckschd.p_disp_detail_sched?term_in=201503&crn_in=22497",
    },
    Object {
      "crn": "19962",
      "info": "Managerial Accounting",
      "meetings": Array [
        Object {
          "endDate": 16561,
          "profs": Array [
            "Christian E. Wurst",
          ],
          "startDate": 16447,
          "times": Object {
            "1": Array [
              Object {
                "end": 70200,
                "start": 61200,
              },
            ],
          },
          "type": "OLL with reqd virtual meetings",
          "where": "Online Room - Place Holder CLSRM",
        },
      ],
      "url": "https://prd-wlssb.temple.edu/prod8/bwckschd.p_disp_detail_sched?term_in=201503&crn_in=19962",
    },
    Object {
      "crn": "24435",
      "info": "Managerial Accounting",
      "meetings": Array [
        Object {
          "endDate": 16547,
          "profs": Array [
            "David M Sigaty",
          ],
          "startDate": 16448,
          "times": Object {
            "2": Array [
              Object {
                "end": 54600,
                "start": 49200,
              },
            ],
            "4": Array [
              Object {
                "end": 54600,
                "start": 49200,
              },
            ],
          },
          "type": "Class",
          "where": "Mita Hall - Japan 00503",
        },
      ],
      "url": "https://prd-wlssb.temple.edu/prod8/bwckschd.p_disp_detail_sched?term_in=201503&crn_in=24435",
    },
  ],
}
`;

exports[`should parse a bunch of deps 1`] = `
Object {
  "classWrapper": Object {
    "deps": Array [],
    "type": "classes",
    "value": Object {
      "classAttributes": Array [],
      "crns": Array [
        "24600",
        "24601",
        "24603",
        "25363",
      ],
    },
  },
  "sectionStartingDatas": Array [
    Object {
      "crn": "24600",
      "info": "Thermodynamic/ Mech",
      "meetings": Array [
        Object {
          "endDate": 16500,
          "profs": Array [
            "Peter J Collings",
          ],
          "startDate": 16454,
          "times": Object {
            "2": Array [
              Object {
                "end": 40200,
                "start": 35700,
              },
            ],
            "4": Array [
              Object {
                "end": 40200,
                "start": 35700,
              },
            ],
          },
          "type": "Class",
          "where": "Science Center L32",
        },
      ],
      "url": "https://myswat.swarthmore.edu/pls/bwckschd.p_disp_detail_sched?term_in=201502&crn_in=24600",
    },
    Object {
      "crn": "24601",
      "info": "Thermodyn/Stat Mechanics - Lab",
      "meetings": Array [
        Object {
          "endDate": 16500,
          "profs": Array [
            "Peter J Collings",
            "MaryAnn Hickman Klassen",
          ],
          "startDate": 16454,
          "times": Object {
            "1": Array [
              Object {
                "end": 58500,
                "start": 47700,
              },
            ],
          },
          "type": "Class",
          "where": "Science Center L44",
        },
      ],
      "url": "https://myswat.swarthmore.edu/pls/bwckschd.p_disp_detail_sched?term_in=201502&crn_in=24601",
    },
    Object {
      "crn": "24603",
      "info": "Thermodyn/Stat Mechanics - Lab",
      "meetings": Array [
        Object {
          "endDate": 16500,
          "profs": Array [
            "MaryAnn Hickman Klassen",
          ],
          "startDate": 16454,
          "times": Object {
            "3": Array [
              Object {
                "end": 58500,
                "start": 47700,
              },
            ],
          },
          "type": "Class",
          "where": "Science Center L44",
        },
      ],
      "url": "https://myswat.swarthmore.edu/pls/bwckschd.p_disp_detail_sched?term_in=201502&crn_in=24603",
    },
    Object {
      "crn": "25363",
      "info": "Thermodyn/Stat Mechanics - Lab",
      "meetings": Array [
        Object {
          "endDate": 16500,
          "profs": Array [
            "Peter J Collings",
          ],
          "startDate": 16454,
          "times": Object {
            "4": Array [
              Object {
                "end": 58500,
                "start": 47700,
              },
            ],
          },
          "type": "Class",
          "where": "Science Center L44",
        },
      ],
      "url": "https://myswat.swarthmore.edu/pls/bwckschd.p_disp_detail_sched?term_in=201502&crn_in=25363",
    },
  ],
}
`;

exports[`should parse a bunch of meetings 1`] = `
Object {
  "classWrapper": Object {
    "deps": Array [],
    "type": "classes",
    "value": Object {
      "classAttributes": Array [],
      "crns": Array [
        "12090",
      ],
    },
  },
  "sectionStartingDatas": Array [
    Object {
      "crn": "12090",
      "info": "The Evolution of U.S. Aerospace Power II",
      "meetings": Array [
        Object {
          "endDate": 16457,
          "profs": Array [
            "Nicholas A Vallera",
          ],
          "startDate": 16457,
          "times": Object {
            "4": Array [
              Object {
                "end": 27600,
                "start": 21600,
              },
            ],
          },
          "type": "Class",
          "where": "TBA",
        },
        Object {
          "endDate": 16471,
          "profs": Array [
            "Nicholas A Vallera",
          ],
          "startDate": 16471,
          "times": Object {
            "4": Array [
              Object {
                "end": 27600,
                "start": 21600,
              },
            ],
          },
          "type": "Class",
          "where": "TBA",
        },
        Object {
          "endDate": 16485,
          "profs": Array [
            "Nicholas A Vallera",
          ],
          "startDate": 16485,
          "times": Object {
            "4": Array [
              Object {
                "end": 27600,
                "start": 21600,
              },
            ],
          },
          "type": "Class",
          "where": "TBA",
        },
        Object {
          "endDate": 16499,
          "profs": Array [
            "Nicholas A Vallera",
          ],
          "startDate": 16499,
          "times": Object {
            "4": Array [
              Object {
                "end": 27600,
                "start": 21600,
              },
            ],
          },
          "type": "Class",
          "where": "TBA",
        },
        Object {
          "endDate": 16513,
          "profs": Array [
            "Nicholas A Vallera",
          ],
          "startDate": 16513,
          "times": Object {
            "4": Array [
              Object {
                "end": 27600,
                "start": 21600,
              },
            ],
          },
          "type": "Class",
          "where": "TBA",
        },
        Object {
          "endDate": 16527,
          "profs": Array [
            "Nicholas A Vallera",
          ],
          "startDate": 16527,
          "times": Object {
            "4": Array [
              Object {
                "end": 27600,
                "start": 21600,
              },
            ],
          },
          "type": "Class",
          "where": "TBA",
        },
        Object {
          "endDate": 16541,
          "profs": Array [
            "Nicholas A Vallera",
          ],
          "startDate": 16541,
          "times": Object {
            "4": Array [
              Object {
                "end": 27600,
                "start": 21600,
              },
            ],
          },
          "type": "Class",
          "where": "TBA",
        },
        Object {
          "endDate": 16555,
          "profs": Array [
            "Nicholas A Vallera",
          ],
          "startDate": 16555,
          "times": Object {
            "4": Array [
              Object {
                "end": 27600,
                "start": 21600,
              },
            ],
          },
          "type": "Class",
          "where": "TBA",
        },
      ],
      "url": "https://prd-wlssb.temple.edu/prod8/bwckschd.p_disp_detail_sched?term_in=201503&crn_in=12090",
    },
  ],
}
`;
